### PR TITLE
Feature/issue 144 multiple builders support

### LINKED ISSUE #144: Multiple Builder pattern support
**Description**

As per today, the `BeanTransformer` is able to transform Java Bean using the following builder pattern:

```
public class ItemType {
    private final Class<?> objectClass;
    private final Class<?> genericClass;

    ItemType(final Class<?> objectClass, final Class<?> genericClass) {
        this.objectClass = objectClass;
        this.genericClass = genericClass;
    }

    public static ItemTypeBuilder builder() {
        return new ItemType.ItemTypeBuilder();
    }

    // getter methods

    public static class ItemTypeBuilder {
        private Class<?> objectClass;
        private Class<?> genericClass;

        ItemTypeBuilder() {
        }

        public ItemTypeBuilder objectClass(final Class<?> objectClass) {
            this.objectClass = objectClass;
            return this;
        }

        public ItemTypeBuilder genericClass(final Class<?> genericClass) {
            this.genericClass = genericClass;
            return this;
        }

        public com.hotels.transformer.model.ItemType build() {
            return new ItemType(this.objectClass, this.genericClass);
        }
    }
}
```

The purpose is to support also the following pattern too:

```
public class ItemType {
    private final Class<?> objectClass;
    private final Class<?> genericClass;

    ItemType(final ItemTypeBuilder builder) {
        this.objectClass = builder.objectClass;
        this.genericClass = builder.genericClass;
    }

    public static ItemTypeBuilder builder() {
        return new ItemType.ItemTypeBuilder();
    }

    // getter methods

    public static class ItemTypeBuilder {
        private Class<?> objectClass;
        private Class<?> genericClass;

        ItemTypeBuilder() {
        }

        public ItemTypeBuilder objectClass(final Class<?> objectClass) {
            this.objectClass = objectClass;
            return this;
        }

        public ItemTypeBuilder genericClass(final Class<?> genericClass) {
            this.genericClass = genericClass;
            return this;
        }

        public com.hotels.transformer.model.ItemType build() {
            return new ItemType(this);
        }
    }
}
```

**Acceptance Criteria**

* **The solution must not degrade the performances**
* The `BeanTransformer` automatically recognized the Builder pattern and use it properly

### PR DESCRIPTION
**Pull request by fborriello**
## Description

Closes #144 
Makes the library able to transform Java Bean using the following pattern too:

```
public class ItemType {
    private final Class<?> objectClass;
    private final Class<?> genericClass;

    ItemType(final ItemTypeBuilder builder) {
        this.objectClass = builder.objectClass;
        this.genericClass = builder.genericClass;
    }

    public static ItemTypeBuilder builder() {
        return new ItemType.ItemTypeBuilder();
    }

    // getter methods

    public static class ItemTypeBuilder {
        private Class<?> objectClass;
        private Class<?> genericClass;

        ItemTypeBuilder() {
        }

        public ItemTypeBuilder objectClass(final Class<?> objectClass) {
            this.objectClass = objectClass;
            return this;
        }

        public ItemTypeBuilder genericClass(final Class<?> genericClass) {
            this.genericClass = genericClass;
            return this;
        }

        public com.hotels.transformer.model.ItemType build() {
            return new ItemType(this);
        }
    }
}
```
Fixes #144 

## How it works?

Java Bean with the Standard Builder pattern are transformed natively, but to enable it for Java Beans using a **custom Builder pattern**, It's needed to enable the feature as following:

```
ToBean toBean = new BeanTransformer()
                         .setCustomBuilderTransformationEnabled(true)
                         .transform(sourceObject, ToBean.class);
```

## How Has This Been Tested?

Please describe the tests that you ran to verify your changes. Please also note any relevant details for your test configuration.

- [X] It transforms Java Bean using the above builder pattern
- [X] It transforms Java Bean using the standard builder pattern
- [X] All the existing feature are working fine

## Checklist:

- [X] The branch follows the best practices naming convention:
     - Use grouping tokens (words) at the beginning of your branch names.
        * `feature`: Feature I'm adding or expanding
        * `bug`: Bug fix or experiment
        * `wip`: Work in progress; stuff I know won't be finished soon
- [X] My code follows the style guidelines of this project
- [X] I have performed a self-review of my own code
- [X] I have commented on my code, particularly in hard-to-understand areas
- [X] I have made corresponding changes to the documentation
- [X] My changes generate no new warnings
- [X] Any dependent changes have been merged and published in downstream modules
- [X] My changes have no bad impacts on performances
- [X] My changes have been tested through Unit Tests and the overall coverage has not decreased. Check the coverall report for your branch: [here](https://coveralls.io/github/HotelsDotCom/bull)
- [X] Any implemented change has been added in the [CHANGELOG.md](./CHANGELOG.md) file 
- [X] Any implemented change has been added in the [CHANGELOG-JDK8.md](./CHANGELOG-JDK8.md) file 
- [x] My changes have been applied on a separate branch too with compatibility with Java 8. Follow this [instructions](https://github.com/HotelsDotCom/bull/blob/master/RELEASE.md#3-prepare-the-jdk8-release) for help.
- [X] The maven version has been increased.

### CODE REVIEW

File: bull-common/src/test/java/com/hotels/transformer/utils/ClassUtilsTest.java [hunk 1]
```diff
@@ -997,12 +1003,97 @@ public void testGetConcreteClassWorksAsExpected(final String testCaseDescription
...
     }
 
+    /**
+     * Test that the a {@link MissingMethodException} is raised if the Builder class has no {@code build()} method.
+     * @param testCaseDescription the test case description
+     * @param containerTestClass the test class containing the Builder
+     * @param builderClass the Builder nested class
+     */
+    @Test(dataProvider = "dataGetBuilderMethodExceptionTesting", expectedExceptions = MissingMethodException.class)
+    public void testGetBuildMethodThrowsExceptionIfMethodIsMissing(final String testCaseDescription, final Class<?> containerTestClass, final Class<?> builderClass) {
```
Comment (leharb) [inline bull-common/src/test/java/com/hotels/transformer/utils/ClassUtilsTest.java:1013]: The parameter`testCaseDescription` is not being used anywhere

Comment (fborriello) [reply bull-common/src/test/java/com/hotels/transformer/utils/ClassUtilsTest.java:1013]: yes it is, is the first argument passed by the data provider. That description is shown at compile time in the console and in your IDE while running the test. It helps the troubleshooting

File: bull-bean-transformer/src/main/java/com/hotels/beans/transformer/TransformerImpl.java [hunk 2]
```diff
@@ -90,6 +97,34 @@
...
+            try {
+                k = classUtils.getNoArgsConstructor(targetClass).get();
+                injectAllFields(sourceObj, k, breadcrumb);
+            } catch (Exception e) {
+                throw new InvalidBeanException(e.getMessage(), e);
+            }
+        } else {
+            k = injectValues(sourceObj, targetClass, classUtils.getAllArgsConstructor(targetClass), breadcrumb, false);
+            if (classType.is(MIXED)) {
+                injectNotFinalFields(sourceObj, k, breadcrumb);
```
Comment (leharb) [inline bull-bean-transformer/src/main/java/com/hotels/beans/transformer/TransformerImpl.java:122]: Should we be handling an exception here also, like in the above `if` block?

Comment (fborriello) [reply bull-bean-transformer/src/main/java/com/hotels/beans/transformer/TransformerImpl.java:122]: we can avoid handling an exception as it's already managed inside the method and the one throws needs to be raised.
Different is for the previous block, where there is the default constructor invocation (that is a `Supplier`) and the exception thrown is not managed.

**Pull request merged**